packages/SwingSet/src/vats/vattp/vat-vattp.js

```diff
@@ -16,181 +23,255 @@ import { Far } from '@endo/marshal';
+    () => 7n,
+  );
+  // TODO: Should this collection be Weak (non-iterable)?
+  const networkHostNames = provideDurableSetStore(baggage, 'networkHostNames');
+  const networkHostKindHandle = provideKindHandle(baggage, 'networkHost');
+  const initNetworkHostState = (allegedName, comms, console) => {
+    assert(
+      !networkHostNames.has(allegedName),
+      X`already have host for ${allegedName}`,
+    );
+    networkHostNames.add(allegedName);
+    const connectionName = `connection-${nextConnectionNumber}`;
+    nextConnectionNumber += 1n;
+    baggage.set(nextConnectionNumberBaggageKey, nextConnectionNumber);
+    return {
+      comms,
+      console,
+
+      allegedName,
+      connection: null,
+      connectionName,
+      exportedObjects: makeScalarBigMapStore(
+        `network host exported objects by address: ${connectionName} as ${allegedName}`,
+        { durable: true },
+      ),
+      nextAddressNumber: 1n,
+      // This assumes one connection per host.
+      // If that is not the case, receiver should instead be a
+      // durable makeScalarBigWeakMapStore keyed by connection.
+      receiver: null,
+      remoteLocatorUnum: null,
+    };
+  };
+  const makeNetworkHost = defineDurableKindMulti(
+    networkHostKindHandle,
+    initNetworkHostState,
+    {
+      host: {
+        publish: ({ state }, obj) => {
+          const address = `/alleged-name/${state.allegedName}/egress/${state.nextAddressNumber}`;
+          state.nextAddressNumber += 1n;
+          state.exportedObjects.init(address, obj);
+          return address;
+        },
+        lookup: ({ state }, address) => {
+          return E(state.remoteLocatorUnum).lookup(address);
+        },
+      },
+      commsSetReceiver: {
+        setReceiver: ({ state }, receiver) => {
+          assert(!state.receiver, X`setReceiver is call-once`);
+          assert(receiver, X`receiver must not be empty`);
+          state.receiver = receiver;
+        },
+      },
+      commsTransmitter: {
+        transmit: ({ state }, msg) => {
+          // 'msg' will be a string (vats/comms/outbound.js deliverToRemote)
+          E(state.connection).send(msg);
+        },
+      },
+      localLocatorUnum: {
+        lookup: ({ state }, address) => {
+          return state.exportedObjects.get(address);
+        },
+      },
+      handler: {
+        onOpen: async ({ state, facets: self }, connection, ..._args) => {
+          const name = state.connectionName;
+          assert(!state.connection, X`host ${name} already opened`);
+          state.connection = connection;
 
-  return handler;
+          const comms = state.comms;
+          // TODO: Can these calls be made in parallel?
+          await E(comms).addRemote(
+            name,
+            self.commsTransmitter,
+            self.commsSetReceiver,
+          );
+          await E(comms).addEgress(name, 0, self.localLocatorUnum);
+          state.remoteLocatorUnum = await E(comms).addIngress(name, 0);
```

Comment:
One tiny difference between this and the original: `onOpen()` will stall until `comms.addIngress` finishes, whereas the original would proceed along. I think that's fine, `addIngress` is prompt (it doesn't wait for anything interesting) so there's not a risk of deadlock. And I do like the simplification that the `await` provides.. easier to read this way.